File: app/src/main/java/com/phucnguyen/lovereminder/ui/activity/BaseActivity.java
package com.phucnguyen.lovereminder.ui.activity;

import android.content.SharedPreferences;
import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.phucnguyen.lovereminder.R;

public class BaseActivity extends AppCompatActivity {
    private static final String PREF_USER_SETTINGS = "user_prefs";
    public static final String FIELD_THEME = "theme_id";
    private int mCurrentThemeId = R.style.AppThemeBase_Rose;
    private SharedPreferences mSharedPref;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mSharedPref = getSharedPreferences(PREF_USER_SETTINGS, MODE_PRIVATE);
        mCurrentThemeId = mSharedPref.getInt(FIELD_THEME, mCurrentThemeId);
    }

    protected void setTheme() {
        setTheme(mCurrentThemeId);
    }

    public void switchTheme(int themeId) {
        mCurrentThemeId = themeId;
        mSharedPref.edit().putInt(FIELD_THEME, mCurrentThemeId).apply();
        recreate();
    }
}

File: app/src/main/java/com/phucnguyen/lovereminder/ui/activity/BaseActivity.kt
package com.phucnguyen.lovereminder.ui.activity

import android.os.Bundle
import android.util.TypedValue
import android.view.Window
import android.view.WindowManager
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.phucnguyen.lovereminder.R
import com.phucnguyen.lovereminder.viewmodel.BaseActivityViewModel
import kotlinx.coroutines.launch


open class BaseActivity : AppCompatActivity() {
    private val viewModel: BaseActivityViewModel by viewModels()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setTheme(viewModel.getCurrentTheme())

        setStatusBarColor()

        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.themeFlow.collect {
                    if (viewModel.isThemeChanged()) {
                        viewModel.updatePreviousTheme()
                        //call recreate to apply the new theme
                        recreate()
                    }
                }
            }
        }
    }

    private fun setStatusBarColor() {
        val currentTheme = resources.newTheme()
        currentTheme.applyStyle(viewModel.getCurrentTheme(), true)
        val typedValue = TypedValue()
        currentTheme.resolveAttribute(R.attr.colorPrimaryDark, typedValue, true);
        val colorPrimaryDark = typedValue.data

        val window: Window = window

        // clear FLAG_TRANSLUCENT_STATUS flag:
        window.clearFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS)

        // add FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS flag to the window
        window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS)

        // finally change the color
        window.statusBarColor = colorPrimaryDark
    }
}